File: 2024/01/aoc24_1_1.jl
# AoC 2024 - Day 1 - Task 1
# https://erik-skopp.de/AdventofCode/2024/1/
# https://github.com/eskopp/AdventOfCode/blob/main/2024/01/aoc24_1_1.jl
#
# Day 1 - Historian Hysteria
# result: 1722302

# Change the working directory to the specified path
cd("2024/01")

# Unsorted arrays for left and right column values
left_column = []
right_column = []

# Read the file and split data into columns
open("input.in", "r") do file
    for line in eachline(file)
        columns = split(strip(line), "   ") # Split columns based on triple spaces
        push!(left_column, columns[1])
        push!(right_column, columns[2])
    end
end

# Sort the left and right columns
sorted_left = sort(parse.(Int, left_column))
sorted_right = sort(parse.(Int, right_column))

# Calculate absolute differences between paired values
differences = [abs(sorted_left[i] - sorted_right[i]) for i in 1:length(sorted_left)]

# Compute the total sum of the differences
total_difference = sum(differences)

# Output the result
println(total_difference)

File: 2024/01/aoc24_1_1.rs
// AoC 2024 - Day 1 - Task 1
// https://erik-skopp.de/AdventofCode/2024/1/
// https://github.com/eskopp/AdventOfCode/blob/main/2024/01/aoc24_1_1.rs
//
// Day 1 - Historian Hysteria
// result: 1722302

use std::fs::File;
use std::io::{self, BufRead};

fn calculate_result(file_path: &str) -> io::Result<i32> {
    // Datei öffnen
    let file = File::open(file_path)?;
    let reader = io::BufReader::new(file);

    // Unsortierte Listen 
    let mut left: Vec<i32> = Vec::new();
    let mut right: Vec<i32> = Vec::new();

    // split
    for line in reader.lines() {
        let line = line?;
        let columns: Vec<&str> = line.trim().split("   ").collect();
        if let [left_value, right_value] = columns[..] {
            left.push(left_value.parse::<i32>().unwrap());
            right.push(right_value.parse::<i32>().unwrap());
        }
    }

    // Listen sortieren
    left.sort();
    right.sort();

    // Betragsdifferenzen 
    let result: i32 = left
        .iter()
        .zip(right.iter())
        .map(|(l, r)| (l - r).abs())
        .sum();

    Ok(result)
}

fn main() -> io::Result<()> {
    // Pfad zur Eingabedatei
    let file_path = "input.in";

    // Funktion aufrufen und Ergebnis ausgeben
    match calculate_result(file_path) {
        Ok(result) => println!("Ergebnis: {}", result),
        Err(e) => eprintln!("Fehler: {}", e),
    }

    Ok(())
}
